pennylane/templates/subroutines/uccsd_unitary.py

```diff
@@ -0,0 +1,231 @@
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+r"""
+Contains the ``UCCSDUnitary`` template.
+"""
+import numpy as np
+
+import pennylane as qml
+
+# pylint: disable-msg=too-many-branches,too-many-arguments,protected-access
+from pennylane.templates.decorator import template
+from pennylane.templates.subroutines import DoubleExcitationUnitary, SingleExcitationUnitary
+from pennylane.templates.utils import (
+    check_no_variable,
+    check_shape,
+    check_type,
+    check_wires,
+    get_shape,
+)
+
+
+@template
+def UCCSDUnitary(weights, wires, ph=None, pphh=None, init_state=None):
+    r"""Implements the the Unitary Coupled-Cluster Singles and Doubles (UCCSD) ansatz
+    by applying the
+    :func:`SingleExcitationUnitary <pennylane.templates.subroutine.SingleExcitationUnitary>`
+    and :func:`DoubleExcitationUnitary <pennylane.templates.subroutine.DoubleExcitationUnitary>`
+    templates to exponentiate the coupled-cluster excitation operator. UCCSD is a VQE ansatz
+    commonly used to run quantum chemistry simulations.
+
+    The UCCSD unitary, within the first-order Trotter approximation, is given by:
+
+    .. math::
+
+        \hat{U}(\vec{\theta}) =
+        \prod_{p > r} \mathrm{exp} \Big\{\theta_{pr}
+        (\hat{c}_p^\dagger \hat{c}_r-\mathrm{H.c.}) \Big\}
+        \prod_{p > q > r > s} \mathrm{exp} \Big\{\theta_{pqrs}
+        (\hat{c}_p^\dagger \hat{c}_q^\dagger \hat{c}_r \hat{c}_s-\mathrm{H.c.}) \Big\}
+
+    where :math:`\hat{c}` and :math:`\hat{c}^\dagger` are the fermionic annihilation and
+    creation operators and the indices :math:`r, s` and :math:`p, q` run over the occupied and
+    unoccupied molecular orbitals, respectively. Using the `Jordan-Wigner transformation
+    <https://arxiv.org/abs/1208.5986>`_ the UCCSD unitary defined above can be written in terms
+    of Pauli matrices as follows (for more details see
+    `arXiv:1805.04340 <https://arxiv.org/abs/1805.04340>`_):
+
+    .. math::
+
+        \hat{U}(\vec{\theta}) = && \prod_{p > r} \mathrm{exp} \Big\{ \frac{i\theta_{pr}}{2}
+        \bigotimes_{a=r+1}^{p-1} \hat{Z}_a (\hat{Y}_r \hat{X}_p - \mathrm{H.c.}) \Big\} \\
+        && \times \prod_{p > q > r > s} \mathrm{exp} \Big\{ \frac{i\theta_{pqrs}}{8}
+        \bigotimes_{b=s+1}^{r-1} \hat{Z}_b \bigotimes_{a=q+1}^{p-1}
+        \hat{Z}_a (\hat{X}_s \hat{X}_r \hat{Y}_q \hat{X}_p +
+        \hat{Y}_s \hat{X}_r \hat{Y}_q \hat{Y}_p +
+        \hat{X}_s \hat{Y}_r \hat{Y}_q \hat{Y}_p +
+        \hat{X}_s \hat{X}_r \hat{X}_q \hat{Y}_p -
+        \{\mathrm{H.c.}\}) \Big\}.
+
+    Args:
+        weights (array): length ``len(ph)+len(pphh)`` vector containing the parameters
+            :math:`\theta_{pr}` and :math:`\theta_{pqrs}` entering the Z rotation in
+            :func:`SingleExcitationUnitary <pennylane.templates.subroutine.SingleExcitationUnitary>`
```

Comment:
Note that the `~` prefix tells sphinx only to display the function name, not the full import path:
```suggestion
            :func:`~.SingleExcitationUnitary`
```
Makes the docstring easier to read when not rendered :) Same applies for `DoubleExcitationUnitary` below

Reply:
I've added some more details to our documentation guidelines page: https://pennylane.readthedocs.io/en/latest/development/guide/documentation.html

Reply:
hanks @josh146 !